src/CommunityToolkit/Xamarin.CommunityToolkit/Helpers/LocalizedString.shared.cs

```diff
@@ -1,42 +1,40 @@
 ﻿using System;
 using System.ComponentModel;
+using Xamarin.Forms.Internals;
 
 namespace Xamarin.CommunityToolkit.Helpers
 {
 #if !NETSTANDARD1_0
-	public class LocalizedString : INotifyPropertyChanged
+	public class LocalizedString : INotifyPropertyChanged, IDisposable
 	{
+		readonly Func<string> generator;
+		readonly LocalizationResourceManager localizationManager;
+
 		public LocalizedString(Func<string> generator = null)
             : this(LocalizationResourceManager.Current, generator)
         {
         }
 
 		public LocalizedString(LocalizationResourceManager localizationManager, Func<string> generator = null)
 		{
-			Generator = generator;
-			localizationManager.PropertyChanged += (sender, args) => Invalidate();
+			this.localizationManager = localizationManager;
+			this.generator = generator;
+			localizationManager.PropertyChanged += Invalidate;
 		}
 
-		Func<string> generator;
-
-		public Func<string> Generator
-		{
-			get => generator;
-			set
-			{
-				generator = value;
-				Invalidate();
-			}
-		}
-
-		public string Localized => Generator?.Invoke();
+		public string Localized => generator?.Invoke();
 
+		[Preserve(Conditional = true)]
 		public static implicit operator LocalizedString(Func<string> func) => new LocalizedString(func);
 
 		public event PropertyChangedEventHandler PropertyChanged;
 
-		public void Invalidate()
+		void Invalidate(object sender, PropertyChangedEventArgs e)
 			=> PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(null));
+
+		public void Dispose() => localizationManager.PropertyChanged -= Invalidate;
+
+		~LocalizedString() => Dispose();
```

Comment:
If you are using the VS it will provide a code snippet for it, when you press `ctrl + .` But if you want I can setup it later or move this to the singleton pattern. :D

Reply:
I don't quite understand what do you mean by singleton. We need to have ability to create new `LocalizedString`s.
I there something wrong with current implementation of `IDisposable`?

Reply:
Singleton is when we use one instance of that object during all application life-cycle, we do [this here](https://github.com/xamarin/XamarinCommunityToolkit/blob/4595d82a4b7fe383fc4da751d08239b33895642e/src/CommunityToolkit/Xamarin.CommunityToolkit/Helpers/LocalizationResourceManager.shared.cs#L11)

`public static LocalizationResourceManager Current { get; } = new LocalizationResourceManager();`

So we don't expose a constructor, normally we set a private constructor, and expose for users the object instance on that `Current` prop. that way we avoid more than one instance.

Is that make sense for you?

Reply:
I know what singleton is =)
I just don't understand how can we use it here. `LocalisationManager` has only one instance for a reason: it's state needs to be shared throughout the application.

In case of `LoclizedSitring`, each one has its own state (generator). How do you propose to make `LocalizedString` a singleton?

Reply:
I'll take a look at that tomorrow, it's time to sleep for me (:

Reply:
Yeah, it's almost 5AM for me 😄
Good night

Reply:
@pictos Inherited `LocalizedString` from `ObservableObject`. Now it should be garbage collected as soon as view model doesn't have reference to it (since `ObservableObject` uses week event), so destructor will trigger and it'll unsubscribe from `LocalizationResourceManager`. 
Does this look good to you?

Reply:
Actually, Finalizers/deconstructor aren't deterministic, since they are triggered by GC we can't say that will be triggered. So the issue couldn't be a memory leak (but this doesn't invalidate memory leak), but other instances listen to that event and lead to odd behaviors. Said that I want to avoid it if it's possible. But you don't need to change your code right now, let's define what will be the best approach first.

For that can you explain better that `Func<string> Generate`? I feel that I didn't get his intent.

Reply:
Could you please clarify what kind of add behaviors we are trying to avoid? Listening to `localizationManager.PropertyChanged` is valid behavior for `LocalizedString` even we are not using it, since it actually will have different value when `localizationManager.PropertyChanged` is triggered. If object that's not already in use report that its value has changed that shouldn't be a problem.

I did some testing and `LocalizedString` will never be garbage collected until `Dispose` is called since it's subscribed to `localizationManager` (so finalizer is useless). We'll need to subscribe using `WeekReference` or something for `LocalizedString` to be automatically collected when the only reference to it is in `localizationManager`. Not sure how to do this for now.

> For that can you explain better that `Func<string> Generate`? I feel that I didn't get his intent.

`LocalizedString` is intended to be companion of `xct:Translate` extension. `Translate` is used in `Views` and automatically updates it's value when culture is changed, `LocalizedString` does the same thing, but is used in `ViewModels`. The only thing that's changing in `Translate` and `LocalizedString` is value of localized string (since different AppResource is used for different culture). That's why `LocalizedString` accepts only `Func<string>`. There is no point in putting `Func<object>` into it. If you have an object some string properties of which are read from AppResources, only those properties should be `LocalizedString`, not the whole object.

Intended use of `LocalizedString` class is to be placed instead of `string` properties in `ViewModel` if value of those properties is dependent on current culture.

I hope that makes sense. If it's still not clear, why `LocalizedString` accept only `Func<string>`, could you please provide some code sample of when accepting `Func<T>` would be useful?

Reply:
I think I got it. This extension method (inspired by [this article](https://faithlife.codes/blog/2008/08/event_subscription_using_weak_references/)) allows for objects that are subscribed using it to be collected while they are subscribed. I tested it and it works
![image](https://user-images.githubusercontent.com/6609929/104773598-fb164000-577d-11eb-9050-7a70c1d04776.png)

```cs
public static class INotifyPropertyChangedEx
{
	public static void WeakSubscribe<T>(this INotifyPropertyChanged target, T subscriber, Action<T, object, EventArgs> action)
	{
		var weakSubscriber = new WeakReference(subscriber, false);
		target.PropertyChanged += handler;

		void handler(object sender, PropertyChangedEventArgs e)
		{
			var s = (T)weakSubscriber.Target;
			if (s is null)
			{
				target.PropertyChanged -= handler;
				return;
			}

			action.Invoke(s, sender, e);
		}
	}
}
```

Here is constructor of `WeekLocalizedString`

```cs
public WeekLocalizedString(LocalizationResourceManager localizationManager, Func<string> generator = null)
{
    this.generator = generator;
    localizationManager.WeakSubscribe(this, (t, _, _) =>
    {
        Console.WriteLine($"Week \"{t.Localized}\" OnPropertyChanged");
        t.OnPropertyChanged(null);
    });
}
```

Reply:
@pictos Refactored `LocalizedString` to use `WeakSubscribe` (and added test to ensure it is disposed when there are no references to it)

Reply:
@maxkoshevoi I'll take a look by tomorrow... In meantime, could a MarkupExtension solve this issue in a more simple way?

Reply:
BTW thanks for the info and article, I'll read all of those tomorrow. Have a good night :D

Reply:
@pictos Good point about MarkupExtension. I think it could, but I'm not sure. Don't really know what markup extensions capable of

Reply:
After sleeping on it, I don't think you can solve this with `MarkupExtension`. There are two reasons why strings can be provided from the ViewModel, instead on hardcoded into the View: 
1) They are dynamic. Different strings are used depending on the state of the view. This maybe can be solved by `MarkupExtension` that takes resource name from ViewModel and fetches its localized value.
2) And most common reason is that strings are generated via some logic. This cannot be solved with `MarkupExtension` since `MarkupExtension` will have how way to regenerate string with values from new culture.